llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll

```diff
@@ -0,0 +1,345 @@
+; CHECK-NEXT:    addi a1, a1, -1
+; CHECK-NEXT:    or a2, a2, a1
+; CHECK-NEXT:    or a3, a3, a1
+; CHECK-NEXT:    or a1, a4, a1
+; CHECK-NEXT:    sw a2, 0(a0)
+; CHECK-NEXT:    sw a3, 4(a0)
+; CHECK-NEXT:    sw a1, 8(a0)
+; CHECK-NEXT:    ret
+  %1 = load i32, ptr %a, align 4
+  %or = or i32 %1, 65535
+  store i32 %or, ptr %a, align 4
+  %arrayidx.1 = getelementptr inbounds nuw i8, ptr %a, i64 4
+  %2 = load i32, ptr %arrayidx.1, align 4
+  %or.1 = or i32 %2, 65535
+  store i32 %or.1, ptr %arrayidx.1, align 4
+  %arrayidx.2 = getelementptr inbounds nuw i8, ptr %a, i64 8
+  %3 = load i32, ptr %arrayidx.2, align 4
+  %or.2 = or i32 %3, 65535
+  store i32 %or.2, ptr %arrayidx.2, align 4
+  ret void
+}
+
+define i32 @andlow16(i32 %x) {
+; CHECK-LABEL: andlow16:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    zext.h a0, a0
+; CHECK-NEXT:    ret
+  %and = and i32 %x, 65535
+  ret i32 %and
+}
+
+define i32 @andlow24(i32 %x) {
+; RV32-LABEL: andlow24:
+; RV32:       # %bb.0:
+; RV32-NEXT:    slli a0, a0, 8
+; RV32-NEXT:    srli a0, a0, 8
+; RV32-NEXT:    ret
+;
+; RV64-LABEL: andlow24:
+; RV64:       # %bb.0:
+; RV64-NEXT:    slli a0, a0, 40
+; RV64-NEXT:    srli a0, a0, 40
+; RV64-NEXT:    ret
+  %and = and i32 %x, 16777215
+  ret i32 %and
+}
+
+define i32 @compl(i32 %x) {
+; CHECK-LABEL: compl:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    not a0, a0
+; CHECK-NEXT:    ret
+  %not = xor i32 %x, -1
+  ret i32 %not
+}
+
+define i32 @orlow12(i32 %x) {
+; NOZBS32-LABEL: orlow12:
+; NOZBS32:       # %bb.0:
+; NOZBS32-NEXT:    lui a1, 1
+; NOZBS32-NEXT:    addi a1, a1, -1
+; NOZBS32-NEXT:    or a0, a0, a1
+; NOZBS32-NEXT:    ret
+;
+; NOZBS64-LABEL: orlow12:
+; NOZBS64:       # %bb.0:
+; NOZBS64-NEXT:    lui a1, 1
+; NOZBS64-NEXT:    addiw a1, a1, -1
+; NOZBS64-NEXT:    or a0, a0, a1
+; NOZBS64-NEXT:    ret
+;
+; ZBS-LABEL: orlow12:
+; ZBS:       # %bb.0:
+; ZBS-NEXT:    ori a0, a0, 2047
+; ZBS-NEXT:    bseti a0, a0, 11
+; ZBS-NEXT:    ret
+  %or = or i32 %x, 4095
+  ret i32 %or
+}
+
+define i32 @xorlow12(i32 %x) {
+; NOZBS32-LABEL: xorlow12:
+; NOZBS32:       # %bb.0:
+; NOZBS32-NEXT:    lui a1, 1
+; NOZBS32-NEXT:    addi a1, a1, -1
+; NOZBS32-NEXT:    xor a0, a0, a1
+; NOZBS32-NEXT:    ret
+;
+; NOZBS64-LABEL: xorlow12:
+; NOZBS64:       # %bb.0:
+; NOZBS64-NEXT:    lui a1, 1
+; NOZBS64-NEXT:    addiw a1, a1, -1
+; NOZBS64-NEXT:    xor a0, a0, a1
+; NOZBS64-NEXT:    ret
+;
+; ZBS-LABEL: xorlow12:
+; ZBS:       # %bb.0:
+; ZBS-NEXT:    xori a0, a0, 2047
+; ZBS-NEXT:    binvi a0, a0, 11
+; ZBS-NEXT:    ret
+  %xor = xor i32 %x, 4095
+  ret i32 %xor
+}
+
+define i64 @andimm64(i64 %x) {
+; RV32-LABEL: andimm64:
+; RV32:       # %bb.0:
+; RV32-NEXT:    lui a1, 1044496
+; RV32-NEXT:    addi a1, a1, -1
+; RV32-NEXT:    and a0, a0, a1
+; RV32-NEXT:    li a1, 0
+; RV32-NEXT:    ret
+;
+; RV64-LABEL: andimm64:
+; RV64:       # %bb.0:
+; RV64-NEXT:    lui a1, 65281
+; RV64-NEXT:    slli a1, a1, 4
+; RV64-NEXT:    addi a1, a1, -1
+; RV64-NEXT:    and a0, a0, a1
```

Comment:
Here it's possible to optimize a non-int32. Getting the condition right in general seems tricky. Can I call `RISCVMatInt` for `Imm` and `~Imm` to see which one is shorter?

Reply:
This wouldn't be the only place to compare two `getIntMatCost`s, so it should be fine, as long as you check the simpler exclusionary conditions first - maybe even including your `isBitwiseLogicOp` loop.